pallets/ajuna-tournament/src/lib.rs

```diff
@@ -0,0 +1,571 @@
+		TournamentEndingTooEarly,
+		/// An error occurred trying to rank an entity,
+		FailedToRankEntity,
+		/// Tournament configuration is invalid.
+		InvalidTournamentConfig,
+	}
+
+	impl<T: Config<I>, I: 'static> Pallet<T, I> {
+		/// The account ID of the treasury.
+		pub fn global_treasury_account_id() -> AccountIdFor<T> {
+			T::PalletId::get().into_account_truncating()
+		}
+
+		/// The account ID of the sub-account for a given season_id/tournament_id.
+		pub fn tournament_treasury_account_id(
+			season_id: T::SeasonId,
+			tournament_id: TournamentId,
+		) -> T::AccountId {
+			let account = TournamentTreasuryAccount::<T::SeasonId>::new(
+				T::PalletId::get(),
+				season_id,
+				tournament_id,
+			);
+			account.into_account_truncating()
+		}
+
+		fn ensure_valid_tournament(
+			config: &TournamentConfigFor<T, I>,
+		) -> Result<u16, DispatchError> {
+			ensure!(config.start < config.end, Error::<T, I>::InvalidTournamentConfig);
+
+			if let Some(initial_reward) = config.initial_reward {
+				ensure!(initial_reward > 0_u32.into(), Error::<T, I>::InvalidTournamentConfig);
+			}
+
+			if let Some(max_reward) = config.max_reward {
+				ensure!(max_reward > 0_u32.into(), Error::<T, I>::InvalidTournamentConfig);
+			}
+
+			if let Some(fee_perc) = config.take_fee_percentage {
+				ensure!(fee_perc <= 100, Error::<T, I>::InvalidTournamentConfig);
+			}
+
+			let reward_table_total_dist =
+				config.reward_table.iter().fold(0_u16, |a, b| a + (*b as u16));
+			ensure!(reward_table_total_dist <= 100, Error::<T, I>::InvalidTournamentConfig);
+
+			ensure!(
+				config.max_players > 0 && config.max_players <= MAX_PLAYERS,
+				Error::<T, I>::InvalidTournamentConfig
+			);
+
+			Ok(reward_table_total_dist)
+		}
+
+		fn try_get_current_tournament_id_for(
+			season_id: &T::SeasonId,
+		) -> Result<TournamentId, DispatchError> {
+			ActiveTournaments::<T, I>::get(season_id)
+				.ok_or_else(|| Error::<T, I>::NoActiveTournamentForSeason.into())
+		}
+
+		fn process_rank_fee_deposit(
+			payer: &AccountIdFor<T>,
+			fee_percentage: u8,
+		) -> Result<BalanceOf<T, I>, DispatchError> {
+			let base_deposit = T::RankDeposit::get();
+			let fee_deposit = base_deposit
+				.saturating_mul(fee_percentage.into())
+				.checked_div(&100_u32.into())
+				.unwrap_or_default();
+			let global_treasury_account = Self::global_treasury_account_id();
+
+			T::Currency::transfer(
+				payer,
+				&global_treasury_account,
+				fee_deposit,
+				ExistenceRequirement::KeepAlive,
+			)?;
+
+			Ok(base_deposit.saturating_sub(fee_deposit))
+		}
+
+		fn process_max_reward_payout(
+			payer: &AccountIdFor<T>,
+			tournament_treasury_account: &AccountIdFor<T>,
+			rank_deposit: BalanceOf<T, I>,
+			max_reward: BalanceOf<T, I>,
+		) -> DispatchResult {
+			let tournament_treasury_balance =
+				T::Currency::free_balance(tournament_treasury_account);
+			let tournament_deposit_left = max_reward.saturating_sub(tournament_treasury_balance);
```

Comment:
This is done so that we can check how close we are the maximum reward that the account can hold.

So for example if the result of the subtraction shows us that we have less than the rank_deposit left to fill the account, the rest of it goes to the global tournament account instead.

Reply:
Thank you, this helps a lot. It also helps me checking that the code does, what it is supposed to do. :)